### DISCUSSION TITLE
Unable to resolve values from FileSystemDirectoryHandle entries and keys methods

### DISCUSSION THREAD

**Discussion by ardouglas** [wasm-bindgen/wasm-bindgen#4337] [Q&A]
### Describe the Bug

Hello, I'm trying to use the FileSystemDirectoryHandle methods for accessing entries or keys and I'm not having much success.  In my code I've tried tried accessing the values from the async iterators basically as described here (https://github.com/rustwasm/wasm-bindgen/discussions/4054) but I never seem to get the desired objects back from the promises.  

### Steps to Reproduce

Here is the code I'm using.  In this instance it is inside of a web worker but I see the same behavior when running on the main thread.
``` rust 
use js_sys::{wasm_bindgen::prelude::Closure, Array};
use wasm_bindgen::JsCast;
use wasm_bindgen_futures::JsFuture;
use web_sys::{
    wasm_bindgen::JsValue, DedicatedWorkerGlobalScope, FileSystemDirectoryHandle, MessageEvent,
};

fn main() {
    wasm_logger::init(wasm_logger::Config::default());
    console_error_panic_hook::set_once();

    let scope = DedicatedWorkerGlobalScope::from(JsValue::from(js_sys::global()));

    let onmessage = Closure::wrap(Box::new(move |msg: MessageEvent| {
        let handle = msg.data().dyn_into::<FileSystemDirectoryHandle>().unwrap();

        let entries = handle.entries();

        wasm_bindgen_futures::spawn_local(async move {
            while let Ok(val) = entries.next() {
                let obj = JsFuture::from(val).await.unwrap();
                log::info!("{:?}", obj);
            }
        });
    }) as Box<dyn Fn(MessageEvent)>);
    scope.set_onmessage(Some(onmessage.as_ref().unchecked_ref()));
    onmessage.forget();
    scope
        .post_message(&Array::new().into())
        .expect("posting ready message succeeds");
}
```

If applicable, add a link to a test case (as a zip file or link to a repository we can clone).

### Expected Behavior

I would expect to get back a JsValue that has the key, value pair described by the MDN docs at https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/entries

### Actual Behavior
The JsValue returned either has the form:
`JsValue(Object({"done":true}))`
or 
`JsValue(Object({"value":undefined}))`

### Additional Context

I realize it's possible that I'm completely misunderstanding how the API is supposed to work, so any help would be greatly appreciated!

Edited to add:
I have tried js casting the object to various forms. When trying to cast to a Map, naturally it fails, but the value printed in the stack trace _does_ seem to contain a key and a value resembling the expected object, but the done property is marked as false.

**Comment by daxpedda** [comment]
The way I usually figure these things out is to start by writing up a JS example and then translating it to `web-sys`.

I'm happy to look into this for you if you can provide me with an equal working JS example that I would use to debug this.

**Reply by ardouglas** [reply]
Thank-you!  I will make a repo.  Is using trunk as the build tool acceptable or would you prefer a script instead?

**Reply by daxpedda** [reply]
I was just expecting a small JS snippet :smile: 
But yeah, Trunk is fine!

**Reply by ardouglas** [reply]
I've included both a working js example and the non working rust example in this repo https://github.com/ardouglas/wasm-directory-drop  -- please let me know if there is anything I can clarify or if you see what I'm doing wrong!  Thank-you again!

**Reply by ardouglas** [reply]
Whoops, I think I finally understand where I went awry.  I should have read the documentation of the `next` method of https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.AsyncIterator.html more closely.  I did not realize that a simple while let would not terminate the loop naturally -- I was getting the objects, they were just hidden by the thousands of lines of console output.  Thank you very much for your time and help!

**Reply by daxpedda** [reply]
Glad you figured it out!
